### ISSUE TITLE
Convert to a real node_module, deprecate `cypress-cli`, and use Github Releases

### ISSUE THREAD

**Issue by brian-mann** [cypress-io/cypress#316]
It is possible for us to no longer split up the binary from the CLI tool; thus ending the reign of `npm install cypress-cli`.

Instead we should just publish the `cypress` node_module and as part of the `postinstall` hook it could download the actual Cypress binary, just like how `Electron` works.

This would fix a few things:

1. You would no longer have a separate `cypress install` step (that command would go away)
2. Versioning would be done completely by npm convention because `cypress` could be checked into `package.json`
3. We could enforce that all users of a project have the same cypress version, and prevent booting a project if we see the cypress version installed does not match `package.json`
4. This removes the concept of a global Cypress install (which won't work well when multiple projects need different Cypress version)
5. CI scripts would be simplified
6. We wouldn't need to publish `cypress-cli` separately, instead this could actually be the `cypress` node_module
7. We could further extend the CLI and enable it to be an actual requireable node_module
8. We could publish binaries and do proper Github Releases.

Users could still go to Github Releases and download a `tar` or `zip` of Cypress as a standalone app outside of `node_modules`. They could do this to try out Cypress on non-node projects.

**Considerations**
The biggest problem I see is that the Desktop GUI's UI is designed to expect multiple projects, whereas this would move Cypress to be installed once per project.

Cypress also writes global app-data that is irrespective of its installed location. This means it behaves as both a locally installed application and a globally installed application. We would then have to be *very* careful not to have multiple versions obliterate settings or migrations. Instead these would have to co-exist together without problems.

OSX Sierra has a "translocation" feature that really screws us when applications are installed outside of `/Applications`. This could potentially be very problematic.

By suggesting users install Cypress via `node_modules` we are further excluding ourselves to development communities outside of `node` but because all tests in Cypress will *always* be written in  Javascript and because we've baked in a lot of natural support for `node_modules` I don't see this as being a problem.

This could also potentially enable us to remove the concept of **in-app** updates. Instead while we could notify users of updates to Cypress, the actual versioning would take place in `cypress.json`.

To take it one step further, we could detect how Cypress is installed and show different messages. If installed globally or via a standalone download, we could link to the Github Releases. If installed via `package.json` we could automatically update the version akin to `npm install --save XXX`.

**Comment by brian-mann** [comment]
Also related to:

- Dependency Failures https://github.com/cypress-io/cypress/issues/317
- XVFB Permissions https://github.com/cypress-io/cypress/issues/330
- Monorepo https://github.com/cypress-io/cypress/issues/256

**Comment by brian-mann** [comment]
Some thoughts on the `node_modules` interface...

```javascript
const Promise = require('bluebird')
const cypress = require('cypress')

// open with optional options
cypress.open({...})

// run with optional options
cypress.run({...})
.then((results) => {
  // results are
 // {
 //    code: numFailures,
 //    failingTests: [],
 //    screenshots: [],
 //    video: Boolean
 // }
})

// run + record
cypress.run({
  record: true
})

// multi instance to single run
cypress.parallel()
.then((runId) => {
  Promise.all([
    cypress.run({runId: runId, spec: 'spec1'}),
    cypress.run({runId: runId, spec: 'spec2'}),
    cypress.run({runId: runId, spec: 'spec3'}),
  ])
})
```

**Comment by bahmutov** [comment]
Example repo with using Cypress via its NPM module API in https://github.com/cypress-io/cypress-example-module-api There are some bugs, but we can work on these and on API general architecture after 0.20.0

**Comment by jennifer-shehane** [comment]
The code for this is done, but this has yet to be released. We'll update the issue and reference the changelog when it's released.

**Comment by brian-mann** [comment]
Fixed in [`0.20.0`](https://docs.cypress.io/guides/references/changelog.html#0-20-0)
